File: packages/sdks/go/aamp/protocol.go
package aamp

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"mime"
	"net/url"
	"regexp"
	"strings"
)

const (
	AAMPProtocolVersion = "1.1"
	HeaderVersion       = "X-AAMP-Version"
	HeaderIntent        = "X-AAMP-Intent"
	HeaderTaskID        = "X-AAMP-TaskId"
	HeaderDispatchCtx   = "X-AAMP-Dispatch-Context"
	HeaderPriority      = "X-AAMP-Priority"
	HeaderExpiresAt     = "X-AAMP-Expires-At"
	HeaderStatus        = "X-AAMP-Status"
	HeaderErrorMsg      = "X-AAMP-ErrorMsg"
	HeaderStructured    = "X-AAMP-StructuredResult"
	HeaderSuggested     = "X-AAMP-SuggestedOptions"
	HeaderStreamID      = "X-AAMP-Stream-Id"
	HeaderParentTaskID  = "X-AAMP-ParentTaskId"
	HeaderCardSummary   = "X-AAMP-Card-Summary"
	HeaderSessionKey    = "X-AAMP-Session-Key"
	HeaderPairCode      = "X-AAMP-Pair-Code"
	HeaderDispatchRules = "X-AAMP-Dispatch-Context-Rules"
)

var dispatchContextKeyRE = regexp.MustCompile(`^[a-z0-9_-]+$`)

func NormalizeHeaders(headers map[string]string) map[string]string {
	normalized := make(map[string]string, len(headers))
	for key, value := range headers {
		normalized[strings.ToLower(key)] = value
	}
	return normalized
}

func ParseDispatchContextHeader(value string) map[string]string {
	if strings.TrimSpace(value) == "" {
		return nil
	}

	context := map[string]string{}
	for _, part := range strings.Split(value, ";") {
		segment := strings.TrimSpace(part)
		if segment == "" {
			continue
		}
		pieces := strings.SplitN(segment, "=", 2)
		if len(pieces) != 2 {
			continue
		}
		key := strings.ToLower(strings.TrimSpace(pieces[0]))
		if !dispatchContextKeyRE.MatchString(key) {
			continue
		}
		decoded, err := url.QueryUnescape(strings.TrimSpace(pieces[1]))
		if err != nil {
			context[key] = strings.TrimSpace(pieces[1])
			continue
		}
		context[key] = decoded
	}
	if len(context) == 0 {
		return nil
	}
	return context
}

func SerializeDispatchContextHeader(context map[string]string) string {
	if len(context) == 0 {
		return ""
	}
	parts := make([]string, 0, len(context))
	for rawKey, rawValue := range context {
		key := strings.ToLower(strings.TrimSpace(rawKey))
		value := strings.TrimSpace(rawValue)
		if value == "" || !dispatchContextKeyRE.MatchString(key) {
			continue
		}
		parts = append(parts, key+"="+url.QueryEscape(value))
	}
	return strings.Join(parts, "; ")
}

func BuildDispatchHeaders(taskID, priority, expiresAt, sessionKey string, dispatchContext map[string]string, parentTaskID string) map[string]string {
	headers := map[string]string{
		HeaderVersion: AAMPProtocolVersion,
		HeaderIntent:  "task.dispatch",
		HeaderTaskID:  taskID,
		HeaderPriority: func() string {
			if priority == "" {
				return "normal"
			}
			return priority
		}(),
	}
	if expiresAt != "" {
		headers[HeaderExpiresAt] = expiresAt
	}
	if trimmed := strings.TrimSpace(sessionKey); trimmed != "" {
		headers[HeaderSessionKey] = trimmed
	}
	if serialized := SerializeDispatchContextHeader(dispatchContext); serialized != "" {
		headers[HeaderDispatchCtx] = serialized
	}
	if parentTaskID != "" {
		headers[HeaderParentTaskID] = parentTaskID
	}
	return headers
}

func BuildCancelHeaders(taskID string) map[string]string {
	return map[string]string{
		HeaderVersion: AAMPProtocolVersion,
		HeaderIntent:  "task.cancel",
		HeaderTaskID:  taskID,
	}
}

func BuildAckHeaders(taskID string) map[string]string {
	return map[string]string{
		HeaderVersion: AAMPProtocolVersion,
		HeaderIntent:  "task.ack",
		HeaderTaskID:  taskID,
	}
}

func BuildStreamOpenedHeaders(taskID, streamID string) map[string]string {
	return map[string]string{
		HeaderVersion:  AAMPProtocolVersion,
		HeaderIntent:   "task.stream.opened",
		HeaderTaskID:   taskID,
		HeaderStreamID: streamID,
	}
}

func BuildResultHeaders(taskID, status, errorMsg string, structuredResult []StructuredResultField) map[string]string {
	headers := map[string]string{
		HeaderVersion: AAMPProtocolVersion,
		HeaderIntent:  "task.result",
		HeaderTaskID:  taskID,
		HeaderStatus:  status,
	}
	if errorMsg != "" {
		headers[HeaderErrorMsg] = errorMsg
	}
	if encoded, err := encodeStructuredResult(structuredResult); err == nil && encoded != "" {
		headers[HeaderStructured] = encoded
	}
	return headers
}

func BuildHelpHeaders(taskID string, suggestedOptions []string) map[string]string {
	return map[string]string{
		HeaderVersion:   AAMPProtocolVersion,
		HeaderIntent:    "task.help_needed",
		HeaderTaskID:    taskID,
		HeaderSuggested: strings.Join(suggestedOptions, "|"),
	}
}

func BuildCardQueryHeaders(taskID string) map[string]string {
	return map[string]string{
		HeaderVersion: AAMPProtocolVersion,
		HeaderIntent:  "card.query",
		HeaderTaskID:  taskID,
	}
}

func BuildCardResponseHeaders(taskID, summary string) map[string]string {
	return map[string]string{
		HeaderVersion:     AAMPProtocolVersion,
		HeaderIntent:      "card.response",
		HeaderTaskID:      taskID,
		HeaderCardSummary: summary,
	}
}

func validatePairCode(pairCode string) (string, error) {
	trimmed := strings.TrimSpace(pairCode)
	if trimmed == "" {
		return "", fmt.Errorf("pairCode cannot be empty")
	}
	if strings.ContainsAny(trimmed, "\r\n") {
		return "", fmt.Errorf("pairCode contains invalid characters")
	}
	if !isValidBase64URLToken(trimmed) {
		return "", fmt.Errorf("pairCode must be a base64url token")
	}
	return trimmed, nil
}

func isValidBase64URLToken(value string) bool {
	if _, err := base64.RawURLEncoding.DecodeString(value); err == nil {
		return true
	}
	_, err := base64.URLEncoding.DecodeString(value)
	return err == nil
}

func BuildPairRequestHeaders(taskID, pairCode string, dispatchContextRules map[string][]string) (map[string]string, error) {
	validated, err := validatePairCode(pairCode)
	if err != nil {
		return nil, err
	}
	rules := dispatchContextRules
	if rules == nil {
		rules = map[string][]string{}
	}
	return map[string]string{
		HeaderVersion:       AAMPProtocolVersion,
		HeaderIntent:        "pair.request",
		HeaderTaskID:        taskID,
		HeaderPairCode:      validated,
		HeaderDispatchRules: encodeBase64URLJSON(rules),
	}, nil
}

func BuildPairRespondHeaders(taskID string, success bool, reason string) map[string]string {
	status := "rejected"
	if success {
		status = "completed"
	}
	headers := map[string]string{
		HeaderVersion: AAMPProtocolVersion,
		HeaderIntent:  "pair.respond",
		HeaderTaskID:  taskID,
		HeaderStatus:  status,
	}
	if !success {
		if cleaned := sanitizePairReason(reason); cleaned != "" {
			headers[HeaderErrorMsg] = cleaned
		}
	}
	return headers
}

func sanitizePairReason(reason string) string {
	cleaned := strings.TrimSpace(strings.ReplaceAll(strings.ReplaceAll(reason, "\r", " "), "\n", " "))
	return cleaned
}

func encodeBase64URLJSON(value any) string {
	payload, err := json.Marshal(value)
	if err != nil {
		return ""
	}
	return base64.RawURLEncoding.EncodeToString(payload)
}

func decodeBase64URLJSONMapStringSlice(value string) map[string][]string {
	if strings.TrimSpace(value) == "" {
		return map[string][]string{}
	}
	payload, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		// Accept standard base64 with padding, matching Node's decode path.
		normalized := strings.ReplaceAll(strings.ReplaceAll(value, "-", "+"), "_", "/")
		switch len(normalized) % 4 {
		case 2:
			normalized += "=="
		case 3:
			normalized += "="
		}
		payload, err = base64.StdEncoding.DecodeString(normalized)
		if err != nil {
			return map[string][]string{}
		}
	}
	decoded := map[string][]string{}
	if err := json.Unmarshal(payload, &decoded); err != nil {
		return map[string][]string{}
	}
	if decoded == nil {
		return map[string][]string{}
	}
	return decoded
}

func ParseAampHeaders(meta EmailMetadata) (*ParsedMessage, error) {
	headers := NormalizeHeaders(meta.Headers)
	intent := headers[strings.ToLower(HeaderIntent)]
	taskID := headers[strings.ToLower(HeaderTaskID)]
	if intent == "" || taskID == "" {
		return nil, nil
	}

	protocolVersion := headers[strings.ToLower(HeaderVersion)]
	if protocolVersion == "" {
		protocolVersion = AAMPProtocolVersion
	}

	subject, err := decodeSubject(meta.Subject)
	if err != nil {
		subject = meta.Subject
	}

	base := &ParsedMessage{
		ProtocolVersion: protocolVersion,
		Intent:          intent,
		TaskID:          taskID,
		From:            strings.Trim(meta.From, "<>"),
		To:              strings.Trim(meta.To, "<>"),
		MessageID:       meta.MessageID,
		Subject:         subject,
	}

	switch intent {
	case "task.dispatch":
		base.Title = strings.TrimSpace(strings.TrimPrefix(subject, "[AAMP Task]"))
		base.Priority = firstNonEmpty(headers[strings.ToLower(HeaderPriority)], "normal")
		base.ExpiresAt = headers[strings.ToLower(HeaderExpiresAt)]
		base.SessionKey = headers[strings.ToLower(HeaderSessionKey)]
		base.DispatchContext = ParseDispatchContextHeader(headers[strings.ToLower(HeaderDispatchCtx)])
		base.ParentTaskID = headers[strings.ToLower(HeaderParentTaskID)]
		base.BodyText = normalizeBodyText(meta.BodyText)
	case "task.cancel":
		base.BodyText = normalizeBodyText(meta.BodyText)
	case "task.result":
		body := parseTaskResultBody(meta.BodyText)
		base.Status = firstNonEmpty(headers[strings.ToLower(HeaderStatus)], "completed")
		base.Output = body.Output
		base.ErrorMsg = firstNonEmpty(body.ErrorMsg, headers[strings.ToLower(HeaderErrorMsg)])
		base.StructuredResult, _ = decodeStructuredResult(headers[strings.ToLower(HeaderStructured)])
	case "task.help_needed":
		body := parseTaskHelpBody(meta.BodyText)
		base.Question = body.Question
		base.BlockedReason = body.BlockedReason
		base.SuggestedOptions = body.SuggestedOptions
	case "task.ack":
	case "task.stream.opened":
		base.StreamID = headers[strings.ToLower(HeaderStreamID)]
	case "pair.request":
		pairCode := headers[strings.ToLower(HeaderPairCode)]
		if pairCode == "" {
			return nil, nil
		}
		base.PairCode = pairCode
		base.DispatchContextRules = decodeBase64URLJSONMapStringSlice(headers[strings.ToLower(HeaderDispatchRules)])
		base.BodyText = normalizeBodyText(meta.BodyText)
	case "pair.respond":
		rawStatus := headers[strings.ToLower(HeaderStatus)]
		if rawStatus == "completed" {
			base.Status = "completed"
			base.Success = true
		} else {
			base.Status = "rejected"
			base.Success = false
		}
		base.ErrorMsg = headers[strings.ToLower(HeaderErrorMsg)]
		base.BodyText = normalizeBodyText(meta.BodyText)
	case "card.query":
		base.BodyText = normalizeBodyText(meta.BodyText)
	case "card.response":
		base.Summary = headers[strings.ToLower(HeaderCardSummary)]
		base.BodyText = normalizeBodyText(meta.BodyText)
	}

	return base, nil
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}
	return ""
}

func encodeStructuredResult(value []StructuredResultField) (string, error) {
	if len(value) == 0 {
		return "", nil
	}
	payload, err := json.Marshal(value)
	if err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(payload), nil
}

func decodeStructuredResult(value string) (any, error) {
	if strings.TrimSpace(value) == "" {
		return nil, nil
	}
	payload, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		return nil, err
	}
	var decoded any
	if err := json.Unmarshal(payload, &decoded); err != nil {
		return nil, err
	}
	return decoded, nil
}

type resultBody struct {
	Output   string
	ErrorMsg string
}

func parseTaskResultBody(body string) resultBody {
	normalized := normalizeBodyText(body)
	if normalized == "" {
		return resultBody{}
	}
	output := extractBodySection(normalized, "Output", []string{"Error"})
	errorMsg := extractBodySection(normalized, "Error", nil)
	if output != "" || errorMsg != "" {
		return resultBody{Output: output, ErrorMsg: errorMsg}
	}
	return resultBody{Output: normalized}
}

type helpBody struct {
	Question         string
	BlockedReason    string
	SuggestedOptions []string
}

func parseTaskHelpBody(body string) helpBody {
	normalized := normalizeBodyText(body)
	if normalized == "" {
		return helpBody{}
	}
	question := extractBodySection(normalized, "Question", []string{"Blocked reason", "Suggested options"})
	blockedReason := extractBodySection(normalized, "Blocked reason", []string{"Suggested options"})
	suggested := parseSuggestedOptions(extractBodySection(normalized, "Suggested options", nil))
	if question != "" || blockedReason != "" || len(suggested) > 0 {
		return helpBody{Question: question, BlockedReason: blockedReason, SuggestedOptions: suggested}
	}
	return helpBody{Question: normalized}
}

func normalizeBodyText(value string) string {
	return strings.TrimSpace(strings.ReplaceAll(value, "\r\n", "\n"))
}

func extractBodySection(bodyText, label string, nextLabels []string) string {
	if bodyText == "" {
		return ""
	}
	normalized := "\n" + bodyText
	lower := strings.ToLower(normalized)
	prefix := "\n" + strings.ToLower(label) + ":"
	start := strings.Index(lower, prefix)
	if start < 0 {
		return ""
	}
	contentStart := start + len(prefix)
	for contentStart < len(normalized) {
		current := normalized[contentStart]
		if current == ' ' || current == '\t' {
			contentStart++
			continue
		}
		break
	}

	end := len(normalized)
	for _, nextLabel := range nextLabels {
		candidate := strings.Index(lower[contentStart:], "\n"+strings.ToLower(nextLabel)+":")
		if candidate >= 0 {
			absolute := contentStart + candidate
			if absolute < end {
				end = absolute
			}
		}
	}

	return strings.TrimSpace(normalized[contentStart:end])
}

func parseSuggestedOptions(value string) []string {
	if strings.TrimSpace(value) == "" {
		return nil
	}
	re := regexp.MustCompile(`^\s*(?:[-*]|\d+\.)\s*`)
	lines := []string{}
	for _, line := range strings.Split(value, "\n") {
		cleaned := strings.TrimSpace(re.ReplaceAllString(line, ""))
		if cleaned != "" {
			lines = append(lines, cleaned)
		}
	}
	return lines
}

func decodeSubject(value string) (string, error) {
	decoder := new(mime.WordDecoder)
	return decoder.DecodeHeader(value)
}

File: packages/sdks/python/aamp_sdk/protocol.py
"""Protocol helpers for AAMP."""

from __future__ import annotations

import base64
import json
import re
from email.header import decode_header
from typing import Any, Mapping

AAMP_PROTOCOL_VERSION = "1.1"

AAMP_HEADER = {
    "VERSION": "X-AAMP-Version",
    "INTENT": "X-AAMP-Intent",
    "TASK_ID": "X-AAMP-TaskId",
    "DISPATCH_CONTEXT": "X-AAMP-Dispatch-Context",
    "PRIORITY": "X-AAMP-Priority",
    "EXPIRES_AT": "X-AAMP-Expires-At",
    "SESSION_KEY": "X-AAMP-Session-Key",
    "STATUS": "X-AAMP-Status",
    "ERROR_MSG": "X-AAMP-ErrorMsg",
    "STRUCTURED_RESULT": "X-AAMP-StructuredResult",
    "SUGGESTED_OPTIONS": "X-AAMP-SuggestedOptions",
    "STREAM_ID": "X-AAMP-Stream-Id",
    "PARENT_TASK_ID": "X-AAMP-ParentTaskId",
    "CARD_SUMMARY": "X-AAMP-Card-Summary",
    "PAIR_CODE": "X-AAMP-Pair-Code",
    "DISPATCH_RULES": "X-AAMP-Dispatch-Context-Rules",
}

_DISPATCH_CONTEXT_KEY_RE = re.compile(r"^[a-z0-9_-]+$")
_PAIR_CODE_RE = re.compile(r"^[A-Za-z0-9_-]+$")


def normalize_headers(headers: Mapping[str, Any]) -> dict[str, str]:
    normalized: dict[str, str] = {}
    for key, value in headers.items():
        header_value = value[0] if isinstance(value, (list, tuple)) else value
        normalized[str(key).lower()] = str(header_value)
    return normalized


def parse_dispatch_context_header(value: str | None) -> dict[str, str] | None:
    if not value:
        return None

    context: dict[str, str] = {}
    for part in value.split(";"):
        segment = part.strip()
        if not segment or "=" not in segment:
            continue
        raw_key, raw_value = segment.split("=", 1)
        key = raw_key.strip().lower()
        encoded_value = raw_value.strip()
        if not _DISPATCH_CONTEXT_KEY_RE.match(key):
            continue
        try:
            from urllib.parse import unquote

            context[key] = unquote(encoded_value)
        except Exception:
            context[key] = encoded_value

    return context or None


def serialize_dispatch_context_header(context: Mapping[str, Any] | None) -> str | None:
    if not context:
        return None

    from urllib.parse import quote

    parts: list[str] = []
    for raw_key, raw_value in context.items():
        key = str(raw_key).strip().lower()
        value = str(raw_value).strip()
        if not value or not _DISPATCH_CONTEXT_KEY_RE.match(key):
            continue
        parts.append(f"{key}={quote(value, safe='')}")

    return "; ".join(parts) or None


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _b64url_decode(value: str) -> bytes:
    raw = value.strip()
    padding = "=" * (-len(raw) % 4)
    try:
        return base64.urlsafe_b64decode(raw + padding)
    except Exception:
        normalized = raw.replace("-", "+").replace("_", "/")
        padding = "=" * (-len(normalized) % 4)
        return base64.b64decode(normalized + padding)


def _encode_base64url_json(value: Any) -> str:
    payload = json.dumps(value, separators=(",", ":")).encode("utf-8")
    return _b64url_encode(payload)


def _decode_base64url_json_map_string_slice(value: str | None) -> dict[str, list[str]]:
    if not value or not str(value).strip():
        return {}
    try:
        decoded = json.loads(_b64url_decode(str(value).strip()).decode("utf-8"))
    except Exception:
        return {}
    if not isinstance(decoded, dict):
        return {}
    result: dict[str, list[str]] = {}
    for key, values in decoded.items():
        if isinstance(values, list):
            result[str(key)] = [str(item) for item in values]
    return result


def _is_valid_base64url_token(value: str) -> bool:
    if not _PAIR_CODE_RE.match(value):
        return False
    try:
        _b64url_decode(value)
        return True
    except Exception:
        return False


def _validate_pair_code(pair_code: str) -> str:
    trimmed = pair_code.strip()
    if not trimmed:
        raise ValueError("pairCode cannot be empty")
    if "\r" in trimmed or "\n" in trimmed:
        raise ValueError("pairCode contains invalid characters")
    if not _is_valid_base64url_token(trimmed):
        raise ValueError("pairCode must be a base64url token")
    return trimmed


def _sanitize_pair_reason(reason: str | None) -> str:
    if not reason:
        return ""
    return reason.replace("\r", " ").replace("\n", " ").strip()


def _encode_structured_result(value: Any | None) -> str | None:
    if value is None:
        return None
    payload = json.dumps(value, separators=(",", ":")).encode("utf-8")
    return _b64url_encode(payload)


def _decode_structured_result(value: str | None) -> Any | None:
    if not value:
        return None
    decoded = _b64url_decode(value)
    return json.loads(decoded.decode("utf-8"))


def _decode_subject(value: str | None) -> str:
    if not value:
        return ""

    parts: list[str] = []
    for piece, encoding in decode_header(value):
        if isinstance(piece, bytes):
            parts.append(piece.decode(encoding or "utf-8", errors="replace"))
        else:
            parts.append(piece)
    return "".join(parts).strip()


def _normalize_body_text(value: str | None) -> str:
    return (value or "").replace("\r\n", "\n").strip()


def _extract_body_section(body_text: str, label: str, next_labels: list[str]) -> str:
    if not body_text:
        return ""
    next_pattern = "|".join(re.escape(item) for item in next_labels)
    if next_pattern:
        pattern = rf"(?:^|\n){re.escape(label)}:\s*([\s\S]*?)(?=\n(?:{next_pattern}):|$)"
    else:
        pattern = rf"(?:^|\n){re.escape(label)}:\s*([\s\S]*?)$"
    match = re.search(pattern, body_text, flags=re.IGNORECASE)
    return match.group(1).strip() if match else ""


def _parse_suggested_options(block: str) -> list[str]:
    if not block.strip():
        return []
    lines: list[str] = []
    for line in block.splitlines():
        cleaned = re.sub(r"^\s*(?:[-*]|\d+\.)\s*", "", line).strip()
        if cleaned:
            lines.append(cleaned)
    return lines


def _parse_task_result_body(body_text: str | None) -> dict[str, Any]:
    normalized = _normalize_body_text(body_text)
    if not normalized:
        return {"output": ""}

    output = _extract_body_section(normalized, "Output", ["Error"])
    error_msg = _extract_body_section(normalized, "Error", [])

    if output or error_msg:
        payload: dict[str, Any] = {"output": output}
        if error_msg:
            payload["errorMsg"] = error_msg
        return payload

    return {"output": normalized}


def _parse_task_help_body(body_text: str | None) -> dict[str, Any]:
    normalized = _normalize_body_text(body_text)
    if not normalized:
        return {"question": "", "blockedReason": "", "suggestedOptions": []}

    question = _extract_body_section(normalized, "Question", ["Blocked reason", "Suggested options"])
    blocked_reason = _extract_body_section(normalized, "Blocked reason", ["Suggested options"])
    suggested_options = _parse_suggested_options(
        _extract_body_section(normalized, "Suggested options", [])
    )

    if question or blocked_reason or suggested_options:
        return {
            "question": question,
            "blockedReason": blocked_reason,
            "suggestedOptions": suggested_options,
        }

    return {"question": normalized, "blockedReason": "", "suggestedOptions": []}


def build_dispatch_headers(
    task_id: str,
    priority: str | None = None,
    expires_at: str | None = None,
    session_key: str | None = None,
    dispatch_context: Mapping[str, Any] | None = None,
    parent_task_id: str | None = None,
) -> dict[str, str]:
    headers = {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "task.dispatch",
        AAMP_HEADER["TASK_ID"]: task_id,
        AAMP_HEADER["PRIORITY"]: priority or "normal",
    }
    if expires_at:
        headers[AAMP_HEADER["EXPIRES_AT"]] = expires_at
    trimmed_session_key = session_key.strip() if session_key else ""
    if trimmed_session_key:
        headers[AAMP_HEADER["SESSION_KEY"]] = trimmed_session_key
    serialized_context = serialize_dispatch_context_header(dispatch_context)
    if serialized_context:
        headers[AAMP_HEADER["DISPATCH_CONTEXT"]] = serialized_context
    if parent_task_id:
        headers[AAMP_HEADER["PARENT_TASK_ID"]] = parent_task_id
    return headers


def build_cancel_headers(task_id: str) -> dict[str, str]:
    return {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "task.cancel",
        AAMP_HEADER["TASK_ID"]: task_id,
    }


def build_ack_headers(task_id: str) -> dict[str, str]:
    return {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "task.ack",
        AAMP_HEADER["TASK_ID"]: task_id,
    }


def build_stream_opened_headers(task_id: str, stream_id: str) -> dict[str, str]:
    return {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "task.stream.opened",
        AAMP_HEADER["TASK_ID"]: task_id,
        AAMP_HEADER["STREAM_ID"]: stream_id,
    }


def build_result_headers(
    task_id: str,
    status: str,
    output: str,
    error_msg: str | None = None,
    structured_result: Any | None = None,
) -> dict[str, str]:
    headers = {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "task.result",
        AAMP_HEADER["TASK_ID"]: task_id,
        AAMP_HEADER["STATUS"]: status,
    }
    if error_msg:
        headers[AAMP_HEADER["ERROR_MSG"]] = error_msg
    encoded_structured_result = _encode_structured_result(structured_result)
    if encoded_structured_result:
        headers[AAMP_HEADER["STRUCTURED_RESULT"]] = encoded_structured_result
    return headers


def build_help_headers(
    task_id: str,
    question: str,
    blocked_reason: str,
    suggested_options: list[str] | None = None,
) -> dict[str, str]:
    del question, blocked_reason
    return {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "task.help_needed",
        AAMP_HEADER["TASK_ID"]: task_id,
        AAMP_HEADER["SUGGESTED_OPTIONS"]: "|".join(suggested_options or []),
    }


def build_card_query_headers(task_id: str) -> dict[str, str]:
    return {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "card.query",
        AAMP_HEADER["TASK_ID"]: task_id,
    }


def build_card_response_headers(task_id: str, summary: str) -> dict[str, str]:
    return {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "card.response",
        AAMP_HEADER["TASK_ID"]: task_id,
        AAMP_HEADER["CARD_SUMMARY"]: summary,
    }


def build_pair_request_headers(
    task_id: str,
    pair_code: str,
    dispatch_context_rules: Mapping[str, list[str]] | None = None,
) -> dict[str, str]:
    validated_pair_code = _validate_pair_code(pair_code)
    rules = dict(dispatch_context_rules or {})
    return {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "pair.request",
        AAMP_HEADER["TASK_ID"]: task_id,
        AAMP_HEADER["PAIR_CODE"]: validated_pair_code,
        AAMP_HEADER["DISPATCH_RULES"]: _encode_base64url_json(rules),
    }


def build_pair_respond_headers(
    task_id: str,
    *,
    success: bool,
    reason: str | None = None,
) -> dict[str, str]:
    status = "completed" if success else "rejected"
    headers = {
        AAMP_HEADER["VERSION"]: AAMP_PROTOCOL_VERSION,
        AAMP_HEADER["INTENT"]: "pair.respond",
        AAMP_HEADER["TASK_ID"]: task_id,
        AAMP_HEADER["STATUS"]: status,
    }
    if not success:
        cleaned_reason = _sanitize_pair_reason(reason)
        if cleaned_reason:
            headers[AAMP_HEADER["ERROR_MSG"]] = cleaned_reason
    return headers


def parse_aamp_headers(meta: Mapping[str, Any]) -> dict[str, Any] | None:
    headers = normalize_headers(meta.get("headers", {}))
    intent = headers.get(AAMP_HEADER["INTENT"].lower())
    task_id = headers.get(AAMP_HEADER["TASK_ID"].lower())
    protocol_version = headers.get(AAMP_HEADER["VERSION"].lower(), AAMP_PROTOCOL_VERSION)
    if not intent or not task_id:
        return None

    from_value = str(meta.get("from", "")).strip("<>")
    to_value = str(meta.get("to", "")).strip("<>")
    subject = _decode_subject(meta.get("subject"))
    message_id = meta.get("messageId") or meta.get("message_id", "")
    body_text = meta.get("bodyText") or meta.get("body_text", "")

    base = {
        "protocolVersion": protocol_version,
        "intent": intent,
        "taskId": task_id,
        "from": from_value,
        "to": to_value,
        "messageId": message_id,
        "subject": subject,
    }

    if intent == "task.dispatch":
        return {
            **base,
            "title": subject.replace("[AAMP Task]", "").strip() or subject,
            "priority": headers.get(AAMP_HEADER["PRIORITY"].lower(), "normal"),
            "expiresAt": headers.get(AAMP_HEADER["EXPIRES_AT"].lower()),
            "sessionKey": headers.get(AAMP_HEADER["SESSION_KEY"].lower()),
            "dispatchContext": parse_dispatch_context_header(
                headers.get(AAMP_HEADER["DISPATCH_CONTEXT"].lower())
            ),
            "parentTaskId": headers.get(AAMP_HEADER["PARENT_TASK_ID"].lower()),
            "bodyText": _normalize_body_text(body_text),
        }

    if intent == "task.cancel":
        return {**base, "bodyText": _normalize_body_text(body_text)}

    if intent == "task.result":
        payload = _parse_task_result_body(body_text)
        structured_result = _decode_structured_result(
            headers.get(AAMP_HEADER["STRUCTURED_RESULT"].lower())
        )
        return {
            **base,
            "status": headers.get(AAMP_HEADER["STATUS"].lower(), "completed"),
            "output": payload.get("output", ""),
            "errorMsg": payload.get("errorMsg") or headers.get(AAMP_HEADER["ERROR_MSG"].lower()),
            "structuredResult": structured_result,
        }

    if intent == "task.help_needed":
        payload = _parse_task_help_body(body_text)
        return {**base, **payload}

    if intent == "task.ack":
        return base

    if intent == "task.stream.opened":
        return {**base, "streamId": headers.get(AAMP_HEADER["STREAM_ID"].lower(), "")}

    if intent == "pair.request":
        pair_code = headers.get(AAMP_HEADER["PAIR_CODE"].lower(), "")
        if not pair_code:
            return None
        return {
            **base,
            "pairCode": pair_code,
            "dispatchContextRules": _decode_base64url_json_map_string_slice(
                headers.get(AAMP_HEADER["DISPATCH_RULES"].lower())
            ),
            "bodyText": _normalize_body_text(body_text),
        }

    if intent == "pair.respond":
        raw_status = headers.get(AAMP_HEADER["STATUS"].lower(), "")
        success = raw_status == "completed"
        status = "completed" if success else "rejected"
        return {
            **base,
            "status": status,
            "success": success,
            "reason": headers.get(AAMP_HEADER["ERROR_MSG"].lower()),
            "errorMsg": headers.get(AAMP_HEADER["ERROR_MSG"].lower()),
            "bodyText": _normalize_body_text(body_text),
        }

    if intent == "card.query":
        return {**base, "bodyText": _normalize_body_text(body_text)}

    if intent == "card.response":
        return {
            **base,
            "summary": headers.get(AAMP_HEADER["CARD_SUMMARY"].lower(), ""),
            "bodyText": _normalize_body_text(body_text),
        }

    return None
